table/block_based_table_reader.cc

```diff
@@ -250,15 +250,15 @@ class BinarySearchIndexReader : public IndexReader {
 
     if (s.ok()) {
       *index_reader = new BinarySearchIndexReader(
-          comparator, std::move(index_block), ioptions.statistics);
+          icomparator, std::move(index_block), ioptions.statistics);
```

Comment:
I thought about this. I think a better way is to not preserve any comparator or statitics pointer in IndexReader, but always pass them in through NewIterator() and other related functions, so that we'll never need to consider this cache life cycle problem. Other way to solve the problem (which is better) is not to put IndexReader in block cache, but actual index block, and IndexReader object is always kept in TableReader. This is harder to implement though but is a cleaner solution. CC @maysamyabandeh @yiwu-arbug